`sets`: move to generic set

howardjohn opened the pull request:
**Please provide a description of this PR:**



**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

howardjohn committed:
Update sets package to be generic

howardjohn committed:
Move usages over to generic

howardjohn committed:
Drop intset

istio-testing commented:
@howardjohn: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/unit-tests-arm64_istio/1568020442172624896) | true | `/test unit-tests-arm64`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/unit-tests-arm64_istio/1568020442172624896) | true | `/test unit-tests-arm64`
unit-tests_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/unit-tests_istio/1568020442122293248) | true | `/test unit-tests`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/unit-tests-arm64_istio/1568020442172624896) | true | `/test unit-tests-arm64`
unit-tests_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/unit-tests_istio/1568020442122293248) | true | `/test unit-tests`
integ-basic-arm64_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-basic-arm64_istio/1568020442952765440) | true | `/test integ-basic-arm64`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/unit-tests-arm64_istio/1568020442172624896) | true | `/test unit-tests-arm64`
unit-tests_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/unit-tests_istio/1568020442122293248) | true | `/test unit-tests`
integ-basic-arm64_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-basic-arm64_istio/1568020442952765440) | true | `/test integ-basic-arm64`
integ-distroless_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-distroless_istio/1568020442478809088) | true | `/test integ-distroless`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/unit-tests-arm64_istio/1568020442172624896) | true | `/test unit-tests-arm64`
unit-tests_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/unit-tests_istio/1568020442122293248) | true | `/test unit-tests`
integ-basic-arm64_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-basic-arm64_istio/1568020442952765440) | true | `/test integ-basic-arm64`
integ-distroless_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-distroless_istio/1568020442478809088) | true | `/test integ-distroless`
integ-cni_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-cni_istio/1568020442256510976) | true | `/test integ-cni`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/unit-tests-arm64_istio/1568020442172624896) | true | `/test unit-tests-arm64`
unit-tests_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/unit-tests_istio/1568020442122293248) | true | `/test unit-tests`
integ-basic-arm64_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-basic-arm64_istio/1568020442952765440) | true | `/test integ-basic-arm64`
integ-distroless_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-distroless_istio/1568020442478809088) | true | `/test integ-distroless`
integ-cni_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-cni_istio/1568020442256510976) | true | `/test integ-cni`
integ-ipv6_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-ipv6_istio/1568020442533335040) | true | `/test integ-ipv6`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn committed:
fix missing intset

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/unit-tests-arm64_istio/1568020442172624896) | true | `/test unit-tests-arm64`
unit-tests_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/unit-tests_istio/1568020442122293248) | true | `/test unit-tests`
integ-basic-arm64_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-basic-arm64_istio/1568020442952765440) | true | `/test integ-basic-arm64`
integ-distroless_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-distroless_istio/1568020442478809088) | true | `/test integ-distroless`
integ-cni_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-cni_istio/1568020442256510976) | true | `/test integ-cni`
integ-ipv6_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-ipv6_istio/1568020442533335040) | true | `/test integ-ipv6`
integ-telemetry_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-telemetry_istio/1568020442327814144) | true | `/test integ-telemetry`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/unit-tests_istio/1568020442122293248) | true | `/test unit-tests`
integ-basic-arm64_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-basic-arm64_istio/1568020442952765440) | true | `/test integ-basic-arm64`
integ-distroless_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-distroless_istio/1568020442478809088) | true | `/test integ-distroless`
integ-cni_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-cni_istio/1568020442256510976) | true | `/test integ-cni`
integ-ipv6_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-ipv6_istio/1568020442533335040) | true | `/test integ-ipv6`
integ-telemetry_istio | d54dd358a325d08c751b5133edb15c9a35943584 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/integ-telemetry_istio/1568020442327814144) | true | `/test integ-telemetry`
lint_istio | 6d19cfc9906c67170b2f1ca8ec2c1f6135e0627e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/lint_istio/1568021237450412032) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

hzxuzhonghu commented on pkg/util/sets/string.go:
```diff
@@ -14,44 +14,47 @@
 // New creates a new Set with the given items.
-func New(items ...string) Set {
-	s := NewWithLength(len(items))
+func New[T constraints.Ordered](items ...T) Set[T] {
```
This is making the usage more complex, as a fundamental lib , better ux is more important than dedup.

ramaraochavali replied:
I agree to an extent with this. The usage certainly is more complex compared to earlier.

ramaraochavali replied:
I agree to an extent with this. The usage certainly is more complex compared to earlier.

ramaraochavali replied:
I agree to an extent with this. The usage certainly is more complex compared to earlier.

howardjohn replied:
Its _slightly_ more complex for strings - but there are 100s of places where we _should_ be using Sets throughout the code but do not because it isn't generic.

If the concern is `sets.Set` -> `sets.Set[string]` is verbose I can add `type String = Set[string]` and then you can use it like `sets.String`?

ramaraochavali replied:
That seems like a good idea for me for readability.

hzxuzhonghu reviewed (commented):
https://planetscale.com/blog/generics-can-make-your-go-code-slower 

Do we need a benchmark?

howardjohn commented:
I'll add one

On Fri, Sep 9, 2022, 2:05 AM Zhonghu Xu ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> https://planetscale.com/blog/generics-can-make-your-go-code-slower
>
> Do we need a benchmark?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/40887#pullrequestreview-1101956842>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEYGXLPDW66EMUVHWDHED3V5L4VXANCNFSM6AAAAAAQIGSKXE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

howardjohn committed:
rename file

howardjohn committed:
add benchmark

howardjohn commented:
```
name            old time/op    new time/op    delta
Set/insert-8      11.3µs ±84%    10.9µs ± 8%     ~     (p=0.156 n=10+9)
Set/contains-8    23.4ns ±25%    10.6ns ± 6%  -54.62%  (p=0.000 n=10+9)
Set/sorted-8       204µs ±54%      96µs ±66%  -53.07%  (p=0.000 n=10+9)

name            old alloc/op   new alloc/op   delta
Set/insert-8        192B ± 0%      192B ± 0%     ~     (all equal)
Set/contains-8     0.00B          0.00B          ~     (all equal)
Set/sorted-8      10.3kB ± 0%    10.2kB ± 0%   -0.24%  (p=0.000 n=10+9)

name            old allocs/op  new allocs/op  delta
Set/insert-8        2.00 ± 0%      2.00 ± 0%     ~     (all equal)
Set/contains-8      0.00           0.00          ~     (all equal)
Set/sorted-8        2.00 ± 0%      1.00 ± 0%  -50.00%  (p=0.000 n=10+10)
```

I think the `contains` being faster is a fluke, but sorted is actually a bit faster (less alloc). Main operations are identical

howardjohn committed:
Add benchmark and alias

howardjohn committed:
Add String alias

howardjohn committed:
Simplify usages

istio-testing commented:
@howardjohn: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio | f958cabbc299ebe9282874ba076418835b1d30cb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/lint_istio/1568273959370100736) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio | f958cabbc299ebe9282874ba076418835b1d30cb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/lint_istio/1568273959370100736) | true | `/test lint`
gencheck_istio | f958cabbc299ebe9282874ba076418835b1d30cb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/gencheck_istio/1568273960208961536) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn committed:
gen

howardjohn commented:
Ok I made 3 tweaks:

* Added benchmarks, showing same performance before and after
* Added `sets.String` type to avoid `sets.Set[string]`
* Elided the type in `New[string](...)` where possible to be just `New()` like before

ramaraochavali commented on pkg/util/sets/set.go:
```diff
@@ -14,44 +14,49 @@
-type Set map[string]struct{}
+type Set[T constraints.Ordered] map[T]struct{}
+
+type String = Set[string]
```
Can we please do the same for Int and also there are still usages like sets.New[string]?

howardjohn replied:
New[string] cannot be avoided in all cases, only when we have given a hint already - either by predeclaring the return type or by passing on the variadic args 

For int - I didn't think we even have any for int anymore, the one we did use I switch to uint32

howardjohn replied:
I think it's strictly worse and only feels better since generics aren't as common *yet*. Ots more wordy than just `New[string]` and less clear

howardjohn replied:
Do you have any other concerns?

istio-testing commented:
@howardjohn: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | 7d76cb49a956ee18601eb09e71c467cf4e1123c4 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/unit-tests-arm64_istio/1569364737785860096) | true | `/test unit-tests-arm64`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | 7d76cb49a956ee18601eb09e71c467cf4e1123c4 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/unit-tests-arm64_istio/1569364737785860096) | true | `/test unit-tests-arm64`
gencheck_istio | 7d76cb49a956ee18601eb09e71c467cf4e1123c4 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/40887/gencheck_istio/1569364749206949888) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn committed:
gen

howardjohn commented:
/retest-required
